File: Programming/PSD_local_work/zin3/code/visualizer/AlarmVisualizer.java
package visualizer;

import model.TemperatureAlarm;
import org.apache.kafka.clients.consumer.*;
import javax.swing.*;
import java.awt.*;
import java.time.Duration;
import java.text.SimpleDateFormat;
import java.util.*;
import com.fasterxml.jackson.databind.ObjectMapper;

public class AlarmVisualizer {
    private static DefaultListModel<String> listModel = new DefaultListModel<>();
    private static final int MAX_ALARMS = 100;
    
    public static void main(String[] args) {
        // Set up the GUI
        JFrame frame = new JFrame("Temperature Alarm Visualizer");
        frame.setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);
        frame.setSize(600, 400);
        
        JList<String> alarmList = new JList<>(listModel);
        alarmList.setFont(new Font(Font.MONOSPACED, Font.PLAIN, 12));
        
        JScrollPane scrollPane = new JScrollPane(alarmList);
        frame.getContentPane().add(scrollPane, BorderLayout.CENTER);
        
        JPanel statsPanel = new JPanel();
        JLabel statsLabel = new JLabel("No alarms yet");
        statsPanel.add(statsLabel);
        frame.getContentPane().add(statsPanel, BorderLayout.SOUTH);
        
        frame.setVisible(true);
        
        // Set up Kafka consumer in a separate thread
        Thread consumerThread = new Thread(() -> {
            // Set up Kafka consumer properties
            Properties props = new Properties();
            props.put("bootstrap.servers", "localhost:9092");
            props.put("group.id", "alarm-visualizer");
            props.put("key.deserializer", "org.apache.kafka.common.serialization.StringDeserializer");
            props.put("value.deserializer", "org.apache.kafka.common.serialization.StringDeserializer");
            props.put("auto.offset.reset", "latest");
            
            Consumer<String, String> consumer = new KafkaConsumer<>(props);
            consumer.subscribe(Collections.singletonList("Alarm"));
            
            ObjectMapper objectMapper = new ObjectMapper();
            SimpleDateFormat sdf = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss");
            
            Map<String, Integer> thermometerAlarmCount = new HashMap<>();
            
            try {
                while (true) {
                    ConsumerRecords<String, String> records = consumer.poll(Duration.ofMillis(100));
                    for (ConsumerRecord<String, String> record : records) {
                        try {
                            TemperatureAlarm alarm = objectMapper.readValue(record.value(), TemperatureAlarm.class);
                            
                            // Format the timestamp
                            Date date = new Date(alarm.getTimestamp());
                            String formattedDate = sdf.format(date);
                            
                            // Create alarm message
                            String alarmMessage = String.format(
                                "⚠️ ALARM: Thermometer %s reported %.2f°C at %s",
                                alarm.getThermometerId(), alarm.getTemperature(), formattedDate);
                            
                            // Update UI on the Event Dispatch Thread
                            SwingUtilities.invokeLater(() -> {
                                // Add new alarm to the top of the list
                                listModel.add(0, alarmMessage);
                                
                                // Keep list at a reasonable size
                                if (listModel.size() > MAX_ALARMS) {
                                    listModel.remove(MAX_ALARMS);
                                }
                                
                                // Update statistics
                                thermometerAlarmCount.put(
                                    alarm.getThermometerId(), 
                                    thermometerAlarmCount.getOrDefault(alarm.getThermometerId(), 0) + 1
                                );
                                
                                StringBuilder stats = new StringBuilder("Alarm Counts: ");
                                for (Map.Entry<String, Integer> entry : thermometerAlarmCount.entrySet()) {
                                    stats.append(entry.getKey()).append("=").append(entry.getValue()).append(", ");
                                }
                                // Remove trailing comma
                                if (stats.length() > 2) {
                                    stats.setLength(stats.length() - 2);
                                }
                                statsLabel.setText(stats.toString());
                            });
                        } catch (Exception e) {
                            System.err.println("Error processing alarm: " + e.getMessage());
                        }
                    }
                }
            } finally {
                consumer.close();
            }
        });
        
        consumerThread.setDaemon(true);
        consumerThread.start();
    }
}

File: Programming/PSD_local_work/zin3/python/code/visualizer/alarm_visualizer.py
import json
import threading
import tkinter as tk
from tkinter import ttk
from confluent_kafka import Consumer
import sys
import os
from datetime import datetime
from collections import Counter

# Add parent directory to path to import model
sys.path.append(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
from model.temperature_alarm import TemperatureAlarm

class AlarmVisualizer:
    def __init__(self, root):
        self.root = root
        self.root.title("Temperature Alarm Visualizer")
        self.root.geometry("600x400")
        
        # Create the list model for alarms
        self.alarm_list = tk.Listbox(root, font=("Courier", 12))
        self.alarm_list.pack(fill=tk.BOTH, expand=True)
        
        # Create scrollbar
        scrollbar = ttk.Scrollbar(self.alarm_list, orient="vertical", command=self.alarm_list.yview)
        scrollbar.pack(side=tk.RIGHT, fill=tk.Y)
        self.alarm_list.config(yscrollcommand=scrollbar.set)
        
        # Create stats panel
        self.stats_frame = tk.Frame(root)
        self.stats_frame.pack(fill=tk.X, side=tk.BOTTOM)
        
        self.stats_label = tk.Label(self.stats_frame, text="No alarms yet")
        self.stats_label.pack(pady=5)
        
        # Stats tracking
        self.thermometer_alarm_count = Counter()
        self.max_alarms = 100
        
        # Start Kafka consumer in a separate thread
        self.consumer_thread = threading.Thread(target=self.consume_alarms, daemon=True)
        self.consumer_thread.start()
    
    def consume_alarms(self):
        # Set up Kafka consumer with confluent-kafka
        consumer_config = {
            'bootstrap.servers': 'localhost:9092',
            'group.id': 'alarm-visualizer',
            'auto.offset.reset': 'latest'
        }
        consumer = Consumer(consumer_config)
        consumer.subscribe(['Alarm'])
        
        try:
            while True:
                msg = consumer.poll(1.0)
                if msg is None:
                    continue
                if msg.error():
                    print(f"Consumer error: {msg.error()}")
                    continue
                    
                # Parse the alarm message
                alarm_data = json.loads(msg.value().decode('utf-8'))
                alarm = TemperatureAlarm.from_dict(alarm_data)
                
                # Format the timestamp
                timestamp_dt = datetime.fromtimestamp(alarm.timestamp / 1000)
                formatted_date = timestamp_dt.strftime('%Y-%m-%d %H:%M:%S')
                
                # Create alarm message
                alarm_message = f"⚠️ ALARM: Thermometer {alarm.thermometer_id} reported {alarm.temperature:.2f}°C at {formatted_date}"
                
                # Update the UI
                self.root.after(0, self.update_ui, alarm_message, alarm.thermometer_id)
                
        except Exception as e:
            print(f"Error in Kafka consumer: {e}")
        finally:
            consumer.close()
    
    def update_ui(self, alarm_message, thermometer_id):
        # Add new alarm to the top of the list
        self.alarm_list.insert(0, alarm_message)
        
        # Keep list at a reasonable size
        if self.alarm_list.size() > self.max_alarms:
            self.alarm_list.delete(self.max_alarms)
        
        # Update statistics
        self.thermometer_alarm_count[thermometer_id] += 1
        
        # Format stats string
        if self.thermometer_alarm_count:
            stats_parts = [f"{id}={count}" for id, count in self.thermometer_alarm_count.items()]
            stats_text = "Alarm Counts: " + ", ".join(stats_parts)
            self.stats_label.config(text=stats_text)

def main():
    root = tk.Tk()
    app = AlarmVisualizer(root)
    root.mainloop()

if __name__ == "__main__":
    main()
